### ISSUE TITLE
Areas for improvement in deployment workflow

### ISSUE THREAD

**Issue by ryanblock** [arc-repos/architect#151]
After just spending a bit of time improving flow control on `deploy` (#145, now released in `> 3.5.0`), @brianleroux were chatting about how we might further improve that workflow.

Here are a few ideas / notes off the top of my head:

- [Important] Read .arc instead of globbing src/
- Enable text-only output / no progress bar with verbose progress messaging (better for CI/CD console)
- Possibly swap out / rewrite 'progress' module; also doesn't give any feedback on processes that are longer running (like step 3, module installation) where the first results may not report back for 10+ seconds)
- Limit parallel deployments above _n_ tps (probably 15, based on experience)
- Delta deploys! speed game changer. add ability to track the functions that have changed since the last deploy (to staging and/or production), and only deploy those functions (unless an override is called to deploy everything, which would always be called if the contents of `src/shared` changed). could be a bit hairy, but would be dope!
- New change: currently, if any steps in 1-5 fail, it's possible deploys may not abort (depending on where/how it fails). This made more sense in the early days, before we did queuing / parallelization and didn't have all those steps broken up. In the next iteration, if any steps in 1-5 fail, we should fully abort the deployment.
- Nice to have, but prob not necessary rn: any functions that errored due to rate limiting could be queued to retry

---

#### Consolidating from issue #141 
**Describe the bug**
During the deploy workflow, npm-related errors or aborted operations are swallowed. In some cases, Lambdas may be updated that do not include hydrated node_modules directories, causing those functions to completely fail.

**To Reproduce**
Can be difficult to reproduce (i.e. npm is unresponsive or slow). One technique could be to have a present package-lock.json file that is out of sync with the package.json file, which should cause npm ci to abort.

**Expected behavior**
We should fail loudly and often when npm-related issues arise, and abort a Lambda deploy instead of letting an update occur that does not include a properly hydrated node_modules folder.

---

### Notes on deploy operations
- Ordered sequentially
- Parallelizeable == sequence of steps can be run in parallel across all functions to be deployed before deployment operations begin
- Queued == sequence of steps must be queued to accommodate AWS API tps limits

PARALLELIZATION STARTS

1. Verification (arc)
2. Pre-deploy plugins
3. Module installation
4. Copy shared
5. Code bundling

PARALLELIZATION ENDS

6. **[Possibly parallelizeable, currently queued]** Zip up function dirs
  - A solution that occurred to me to enable parallelized function zipping would be to establish a second pre-`UpdateFunctionCode` queue to feed into the `UpdateFunctionCode` queue; let's call it the zip queue
  - Thus, `UpdateFunctionCode` operations could pull its chunks from zip queue
  - This would in theory keep the deploy process from having to
    - a) Zip and hold in memory the entire codebase before step 7, or 
    - b) Zip each fn as part of `UpdateFunctionCode` operation, thus impacting reliability of `UpdateFunctionCode` flow control
  - That said, i think combining 6 + 7 as we do today is totally fine and now reliable on larger projects
  - Nore details on considerations for [parallelizing function zipping operation here](https://github.com/arc-repos/arc-workflows/issues/145#issuecomment-425509068)

QUEUING STARTS
7. `UpdateFunctionCode`
8. Post-deploy plugins
DEPLOY FINISHED

**Comment by filmaj** [comment]
Thoughts on splitting this issue up into separate enhancements? I sort of started that I guess with #187

**Comment by brianleroux** [comment]
super into it. the WIP pr https://github.com/arc-repos/architect/pull/208 should close these:

- [x] Read .arc instead of globbing src/
- [x] Enable text-only output
- [x] Possibly swap out / rewrite 'progress' module; 
- [x] Delta deploys!

**Comment by filmaj** [comment]
Holy shit! nice work!

**Comment by brianleroux** [comment]
fixed in `4.1.1`; to test out the WIP `npm i @architect/architect@rc1`
